pkg/controller/controller.go

```diff
@@ -0,0 +1,198 @@
+package controller
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// maxRetries is the number of times an object will be retried before it is dropped out of the queue.
+	// With the current rate-limiter in use (5ms*2^(maxRetries-1)) the following numbers represent the
+	// sequence of delays between successive queueing of an object.
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
+	maxRetries = 15
+
+	// defaultWorkerThreadiness is the number of goroutines spawned
+	// to service an informer event queue.
+	defaultWorkerThreadiness = 10
+
+	// ClientTimeout specifies the timeout for our calls to the API server for
+	// all client operations
+	ClientTimeout = 2 * time.Second
+)
+
+type SharedControllerInterface interface {
+	SyncHandler(key string) error
+}
+
+type Controller struct {
```

Comment:
The structure laid out here is similar to https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure

tbh, I would just copy/paste that core structure and get over it.  Later if you really do have duplicated logic, you could collapse it.  Practically, in places like kube-controller-manager for instance, collapsing it has never been a priority because it wasn't onerous to maintain.  I wrote that skeleton about 4-5 years ago and its still valid.